src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts

```diff
@@ -110,31 +148,92 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
-				return [];
+				return undefined;
+			}
+			const completions: ITerminalCompletion[] | TerminalCompletionList<ITerminalCompletion> | undefined = await provider.provideCompletions(promptValue, cursorPosition, token);
+			if (!completions) {
+				return undefined;
 			}
-			const completions = await provider.provideCompletions(promptValue, cursorPosition, token);
 			const devModeEnabled = this._configurationService.getValue(TerminalSettingId.DevMode);
-			if (completions) {
-				return completions.map(completion => {
-					if (devModeEnabled && !completion.detail?.includes(provider.id)) {
-						completion.detail = `(${provider.id}) ${completion.detail ?? ''}`;
-					}
-					return completion;
-				});
+			const completionItems = Array.isArray(completions) ? completions : completions.items ?? [];
+
+			const itemsWithModifiedLabels = completionItems.map(completion => {
+				if (devModeEnabled && !completion.detail?.includes(provider.id)) {
+					completion.detail = `(${provider.id}) ${completion.detail ?? ''}`;
+				}
+				return completion;
+			});
+
+			if (Array.isArray(completions)) {
+				return itemsWithModifiedLabels;
+			} else if (completions.resourceRequestConfig) {
+				const resourceCompletions = await this._resolveResources(completions.resourceRequestConfig, promptValue, cursorPosition);
+				if (resourceCompletions) {
+					itemsWithModifiedLabels.push(...resourceCompletions);
+				}
+				return itemsWithModifiedLabels;
 			}
-			return [];
+			return;
 		});
 
 		const results = await Promise.all(completionPromises);
-		results.forEach(completions => completionItems.push(...completions));
+		return results.filter(result => !!result).flat();
 	}
-}
 
+	private async _resolveResources(resourceRequestConfig: TerminalResourceRequestConfig, promptValue: string, cursorPosition: number): Promise<ITerminalCompletion[] | undefined> {
+		const cwd = URI.revive(resourceRequestConfig.cwd);
+		const foldersRequested = resourceRequestConfig.foldersRequested ?? false;
+		const filesRequested = resourceRequestConfig.filesRequested ?? false;
+		if (!cwd || (!foldersRequested && !filesRequested)) {
+			return;
+		}
+
+		const resourceCompletions: ITerminalCompletion[] = [];
+		const fileStat = await this._fileService.resolve(cwd, { resolveSingleChildDescendants: true });
+
+		if (!fileStat || !fileStat?.children) {
+			return;
+		}
+
+		for (const stat of fileStat.children) {
+			let kind: TerminalCompletionItemKind | undefined;
+			if (foldersRequested && stat.isDirectory) {
+				kind = TerminalCompletionItemKind.Folder;
+			}
+			if (filesRequested && !stat.isDirectory && (stat.isFile || stat.resource.scheme === 'file')) {
+				kind = TerminalCompletionItemKind.File;
+			}
+			if (kind === undefined) {
+				continue;
+			}
+			const lastWord = promptValue.substring(0, cursorPosition).split(' ').pop();
+			const lastIndexOfDot = lastWord?.lastIndexOf(' .') ?? -1;
+			const lastIndexOfSlash = lastWord?.lastIndexOf('/') ?? -1;
```

Comment:
Should a terminal completion provider also declare its path separator(s)? This would let the extension decide which is running on the remote in that case so it would get the right OS. We need to know somehow whether to use `/` and `\` or just `/`.

Reply:
don't we also know that info 🤔 `_isWindowsPty`